### ISSUE TITLE
Add visual indicator a post/reply has been broadcasted

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#6232]
Currently, there's no way to know if a new post/reply has already been processed by the Notifier worker. This leads to confusion as to wether we are screaming in the void or not. It is especially crucial for single-user instances (like mine) to know if and when a post has been sent out to the world, because it is the only way it can be read by other people.

To avert this problem, I'd like to have a visual indicator on each posts I write that its related Notifier worker task is either pending, processed or deferred. Instant messaging apps are doing the same using a set of icons for the multiple stages of a message transmission (pending reception by messaging hub, received by messaging hub, received by remote contact, read by remote contact in Signal for example).

Possible states:
- Notifier task pending
- Notifier task deferred
- Notifier task incomplete (if some recipients are unreachable) although I would advise against it for public posts.
- Notifier task complete

**Comment by annando** [comment]
That's complicate. The notifier does call possibly hundreds of ```delivery``` and ```apdelivery``` tasks. And these tasks will normally _never_ be complete, since there always will be some receivers currently not listening.

**Comment by MrPetovan** [comment]
That's why I advised it against public posts, but it would make sense for limited posts.

Otherwise, it just is a matter of setting a flag on the item at the end of the Notifier task.

**Comment by MrPetovan** [comment]
How do I make the new `user-item.notified` field available in `Item::select`? I tried adding it in `Item::fieldList` but the `user-item` fields are only added when `$usermode = true`, which doesn't seem to happen in `conversation_add_children` at all.

**Comment by annando** [comment]
On limited posts it's the same. The notifier always calls the delivery jobs, no matter if public or private.

It would be wrong to have it added to the ```user-item``` since this is no user related setting. I would vote for ```item-delivery-data``` since it is only relevant for content that we created by ourselves.

But like I said: You had to wait for every delivery and apdelivery job to be finished before you could set it.

**Comment by MrPetovan** [comment]
Thanks for the pointer. This means the field should be NULL for all items that the current user didn't send themself.

> You had to wait for every delivery and apdelivery job to be finished before you could set it.

Not necessarily. There is a grace period between the moment you post and the moment it is actually sent through the Notifier worker task during which you know it hasn't been in front of anyone else's eyeballs outside your server. During this grace period, editing is "free", meaning that any change will be propagated to remote hosts even without support for formal edits. Same goes with deletion, if an item is deleted before it is sent outside, then it's as if it never existed, which is useful knowledge.

It doesn't have to be perfect knowledge, just a milestone in the item lifecycle. Once the notifier is done, you know the item has been actually scheduled to be sent to all relevant remote hosts. Depending on the worker frequency, the queue priority and the system load, this delay can vary a lot (5-15 mn observed) and it would be useful to have a visual indicator for it.

**Comment by annando** [comment]
The table ```item-delivery-data``` only does have entries for items that are created by us. This saves space. Concerning the delivery times: At my systems the time between sending and receiving normally is normally less then 30 seconds, often less than that. Only when the system is really under heavy load, it will last some minutes.

**Comment by MrPetovan** [comment]
Lucky you! Maybe I should actually do what you suggested, update the flag with each successful delivery. it wouldn't be a binary flag, just a "delivered 8/10" kind of value.

**Comment by annando** [comment]
I just realized another problem. This counter sounds like a good idea - but there are servers out there that deliver content with > 1,000 posts per minute in several parallel worker queues. it is highly likely that the counter wouldn't be updated correctly then.

**Comment by MrPetovan** [comment]
Even with a ``UPDATE `item-delivery-data` SET `delivered` = `delivered` + 1 WHERE ...`` query ? There wouldn't be a concurrency problem.

**Comment by annando** [comment]
I guess this would be better - but would be a problem in the ongoing task of abstracting all database calls.

**Comment by MrPetovan** [comment]
Yes and no, it's standard SQL anyway, there's no MySQL specific function. And we aren't going to support arbitrary storage for data (like NoSQL, etc...).

**Comment by annando** [comment]
I guess you then will need two integer fields in the ```item-delivery-data``` table? One that is increased in the notifier worker when "delivery" and "apdelivery" are spawned, and one in these worker processes, when they are finished? So ```0/0``` will mean, that the notifier hasn't started yet?

**Comment by MrPetovan** [comment]
The first field will be set in Notifier to the number of expected recipients. The second field will be set to 0 at the same point, and then will be incremented in each separate delivery process.

So:
- 0/0: pending Notifier
- 0/10: Notifier done, pending delivery
- 5/10: Delivery underway
- 10/10: Delivery complete, although I doubt that we'll get a perfect number like this for public items, so I may just use a "Delivery complete" at 1/10 for public items only.

**Comment by annando** [comment]
It really doesn't depend on private or public. The delivery process is the same.

**Comment by MrPetovan** [comment]
Reopened after the develop rebase onto 2019.01 stable, delivery counts aren't set, leaving all the indicators as "Pending".

**Comment by MrPetovan** [comment]
I'm starting to see delivery indicators change, it was probably caused by the backed up workerqueue due to the APDelivery bug I introduced.
